cdap-data-fabric/src/main/java/co/cask/cdap/data2/security/ImpersonationUtils.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.security;
+
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.io.Files;
+import com.google.common.io.InputSupplier;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.twill.filesystem.Location;
+import org.apache.twill.filesystem.LocationFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.UndeclaredThrowableException;
+import java.net.URI;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.FileAttribute;
+import java.nio.file.attribute.PosixFilePermission;
+import java.nio.file.attribute.PosixFilePermissions;
+import java.security.PrivilegedExceptionAction;
+import java.util.Set;
+import java.util.concurrent.Callable;
+
+/**
+ *
+ */
+public final class ImpersonationUtils {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ImpersonationUtils.class);
+
+  private ImpersonationUtils() { }
+
+  /**
+   * Returns the path to a keytab file. If it's not a local file, copies it to local file system first.
+   *
+   * @param locationFactory the location factory to use to retrieve the keytab file, if the file is not already local
+   * @param keytabURI the URI of the keytab file
+   * @param tempDir a directory in which to localize the keytab file, if necessary
+   * @return the path to a local keytab file
+   */
+  public static String localizeKeytab(LocationFactory locationFactory, URI keytabURI, File tempDir) throws IOException {
```

Comment:
done. Downside of this is that both the caller of this method (to make a LocalLocation) and the implementation of this method (to avoid copying/localizing) need to now check if the uri is for a local file.